bpf/rtt_tracker.h

```diff
@@ -1,143 +1,161 @@
         }
-        // This is an ACK packet with valid sequence id so a SYN must
-        // have been sent. We can safely update the reverse flow RTT here.
-        update_reverse_flow_rtt(pkt, seq);
+        return 0;
     }
-    return;
-}
+    return -1;
+ }
 
-static __always_inline void __store_tcp_ts(pkt_info *pkt, struct tcphdr *tcp, flow_seq_id *seq_id) {
-    // store timestamp of syn packets.
-    u32 seq = bpf_ntohl(tcp->seq);
-    fill_flow_seq_id(seq_id, pkt, seq, false);
-    long ret = bpf_map_update_elem(&flow_sequences, seq_id, &pkt->current_ts, BPF_NOEXIST);
-    if (trace_messages && ret != 0) {
-        bpf_printk("err saving flow sequence record %d", ret);
-    }
-    return;
-}
+static inline int calculate_flow_rtt_tcp(struct sock *sk, struct sk_buff *skb) {
+	struct tcp_sock *ts;
+    u16 family, flags = 0;
+    u64 rtt = 0, len;
+    int ret = 0;
+    flow_id id;
+    u8 dscp = 0;
 
-static __always_inline void calculate_flow_rtt_tcp(pkt_info *pkt, u8 direction, void *data_end, flow_seq_id *seq_id) {
-    struct tcphdr *tcp = (struct tcphdr *) pkt->l4_hdr;
-    if ( !tcp || ((void *)tcp + sizeof(*tcp) > data_end) ) {
-        return;
-    }
+    __builtin_memset(&id, 0, sizeof(id));
 
-    /* We calculate RTT for both SYN/SYN+ACK and SYN+ACK/ACK and take the maximum of both.*/
-    if (tcp->syn && tcp->ack) { // SYN ACK Packet
-        __calculate_tcp_rtt(pkt, tcp, seq_id);
-        __store_tcp_ts(pkt, tcp, seq_id);
-    }
-    else if (tcp->ack) {
-        __calculate_tcp_rtt(pkt, tcp, seq_id);
+    id.if_index = BPF_CORE_READ(skb, skb_iif);
+    len = BPF_CORE_READ(skb, len);
+
+    // read L2 info
+    rtt_fill_in_l2(skb, &id);
+
+    family = BPF_CORE_READ(sk, __sk_common.skc_family);
+
+    // read L3 info
+    rtt_fill_in_l3(skb, &id, family, &dscp);
+
+    // read TCP info
+    rtt_fill_in_tcp(skb, &id, &flags);
+
+    // read TCP socket rtt and store it in nanoseconds
+    ts = (struct tcp_sock *)(sk);
+    rtt = BPF_CORE_READ(ts, srtt_us) >> 3;
+    rtt *= 1000u;
+
+    // update flow with rtt info
+    id.direction = INGRESS;
+    ret = rtt_lookup_and_update_flow(&id, flags, rtt);
+    if (ret == 0) {
+        return 0;
     }
-    else if (tcp->syn) {
-        __store_tcp_ts(pkt, tcp, seq_id);
+
+    u64 current_ts = bpf_ktime_get_ns();
+    flow_metrics new_flow = {
+        .packets = 1,
+        .bytes = len,
+        .start_mono_time_ts = current_ts,
+        .end_mono_time_ts = current_ts,
+        .flags = flags,
+        .flow_rtt = rtt,
+        .dscp = dscp,
+    };
+    ret = bpf_map_update_elem(&aggregated_flows, &id, &new_flow, BPF_ANY);
+    if (trace_messages && ret != 0) {
+        bpf_printk("error rtt track creating flow %d\n", ret);
     }
+
+    return 0;
 }
 
-static __always_inline void calculate_flow_rtt(pkt_info *pkt, u8 direction, void *data_end) {
-    flow_seq_id seq_id;
-    __builtin_memset(&seq_id, 0, sizeof(flow_seq_id));
+SEC("fentry/tcp_rcv_established")
```

Comment:
for my own knowledge, what's "fentry/tcp_rcv_established" compared to "kprobe/tcp_rcv_established" ?

Reply:
`kprobe` inserted in the function while `fentry` stands for function entry everytime the function is called it will fire similarly there `fexit`the reason I am using double hook is following BCC implementation to make sure we never miss rtt update